src/vulkan-renderer/wrapper/device.cpp

```diff
@@ -31,83 +33,334 @@ constexpr float DEFAULT_QUEUE_PRIORITY = 1.0f;
+    assert(physical_device);
 
-    VkBool32 presentation_supported = VK_FALSE;
+    VkPhysicalDeviceProperties properties{};
+    vkGetPhysicalDeviceProperties(physical_device, &properties);
 
-    // Query if presentation is supported.
-    if (const auto result = vkGetPhysicalDeviceSurfaceSupportKHR(graphics_card, 0, surface, &presentation_supported);
-        result != VK_SUCCESS) {
-        throw VulkanException("Error: vkGetPhysicalDeviceSurfaceSupportKHR failed!", result);
+    VkPhysicalDeviceMemoryProperties memory_properties{};
+    vkGetPhysicalDeviceMemoryProperties(physical_device, &memory_properties);
+
+    VkPhysicalDeviceFeatures features{};
+    vkGetPhysicalDeviceFeatures(physical_device, &features);
+
+    VkDeviceSize total_device_local = 0;
+    for (std::size_t i = 0; i < memory_properties.memoryHeapCount; i++) {
+        const auto &heap = memory_properties.memoryHeaps[i];
+        if ((heap.flags & VK_MEMORY_HEAP_DEVICE_LOCAL_BIT) != 0) {
+            total_device_local += heap.size;
+        }
     }
 
-    return presentation_supported == VK_TRUE;
-}
+    // Default to true in this case where a surface is not passed (and therefore presentation isn't cared about)
+    VkBool32 presentation_supported = VK_TRUE;
+    if (surface != nullptr) {
+        if (const auto result =
+                vkGetPhysicalDeviceSurfaceSupportKHR(physical_device, 0, surface, &presentation_supported);
+            result != VK_SUCCESS) {
+            throw VulkanException("Error: vkGetPhysicalDeviceSurfaceSupportKHR failed!", result);
+        }
+    }
 
-Device::Device(const wrapper::Instance &instance, const VkSurfaceKHR surface, bool enable_vulkan_debug_markers,
-               bool prefer_distinct_transfer_queue, const std::optional<std::uint32_t> preferred_physical_device_index)
-    : m_surface(surface), m_enable_vulkan_debug_markers(enable_vulkan_debug_markers) {
+    const auto extensions = vk_tools::get_all_physical_device_extension_properties(physical_device);
 
-    const auto selected_gpu =
-        VulkanSettingsDecisionMaker::graphics_card(instance.instance(), surface, preferred_physical_device_index);
+    const bool is_swapchain_supported =
+        surface == nullptr || is_extension_supported(extensions, VK_KHR_SWAPCHAIN_EXTENSION_NAME);
 
-    if (!selected_gpu) {
-        throw std::runtime_error("Error: Could not find suitable graphics card!");
+    return DeviceInfo{
+        .name = properties.deviceName,
+        .physical_device = physical_device,
+        .type = properties.deviceType,
+        .total_device_local = total_device_local,
+        .presentation_supported = presentation_supported == VK_TRUE,
+        .swapchain_supported = is_swapchain_supported,
+        .features = features,
+        .extensions = extensions,
+    };
+}
+
+} // namespace
+
+VkPhysicalDevice Device::pick_best_physical_device(std::vector<DeviceInfo> &&physical_device_infos,
```

Comment:
```suggestion
VkPhysicalDevice Device::get_best_device(std::vector<DeviceInfo> &&physical_device_infos,
```

Reply:
Hmm I am undecided. Yeetari what do you think?

Reply:
at least do **not** have two with the same name :D

Reply:
But it's **overloaded** because we need to have it like that for the unit tests. In the unit tests, we can't have "real" Vulkan handles (pointers). So the unit tests use `get_best_physical_device(std::vector<DeviceInfo> &&physical_device_infos,`.

Reply:
Yeah but the functions do something fundamentally different.
One is connected to the real world and the other is just working in vectors.